Start work on solution for blocking_recv and send in mpsc (issue #2629)

xd009642 opened the pull request:
WIP start just to see if I'm in the right ballpark. Not too sure about the busy waiting, so if there's a preferred solution I'd be happy to know.

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
````diff
@@ -150,6 +150,39 @@ impl<T> Receiver<T> {
+    ///     sync_code.join().unwrap()
+    /// }
+    /// ```
+    pub fn blocking_recv(&mut self) -> Option<T> {
+        loop {
+            let res = match self.try_recv() {
+                Ok(t) => Some(t),
+                Err(TryRecvError::Closed) => None,
+                _ => continue,
+            };
+            return res;
+        }
+    }
````
We don't want to just spin-loop.

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
```diff
@@ -150,6 +152,44 @@ impl<T> Receiver<T> {
+        let mut cx = Context::from_waker(&waker);
+
+        loop {
+            if let Poll::Ready(v) = crate::coop::budget(|| self.poll_recv(&mut cx)) {
+                return Ok(v);
+            }
```
The coop budget is not necessary for this.

xd009642 replied:
Done :+1:

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
````diff
@@ -150,6 +152,44 @@ impl<T> Receiver<T> {
+    /// }
+    /// ```
+    pub fn blocking_recv(&mut self) -> Result<Option<T>, crate::park::ParkError> {
+        if Handle::try_current().is_ok() {
+            panic!("blocking_recv cannot be called from within a runtime");
+        }
````
This is not the way we should check this. It should not panic in `spawn_blocking` or `block_in_place`.

xd009642 replied:
What is the best way to check it? I'm not fully sure from looking at `src/runtime/enter.rs` (I might be looking in the wrong place). 

> It is possible that you will not conditional compilation to turn off the panic when only sync is enabled.

I'm not sure I follow this what you mean by this? It shouldn't be possible to conditionally remove the panic based on the features?

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
```diff
@@ -255,6 +298,34 @@ impl<T> Sender<T> {
+        let mut cx = Context::from_waker(&waker);
+
+        loop {
+            if let Poll::Ready(v) = self.poll_ready(&mut cx) {
```
Seems like we should be able to just poll the async `send` function. That would deduplicate some code.

xd009642 replied:
So that thought crossed my mind, but value would be moved into `send` and then I can't use it when returning an error if `park.park()` fails. Unless there's a better way to get around that?

Darksonn replied:
It seems like we should just panic if `park.park()` fails. That would be a bug in our program.

carllerche replied:
Yes, `park.park()` returning an error is legacy. Unwrap is fine. That said, I commented above that you can just use `enter.block_on`.

xd009642 commented:
Looking at the CI a thing I'm wondering about is the `CachedParkThread`, it might be that `blocking` has to become a dependency of `sync`, alternatively `CachedParkThread` is available when `sync` is or another way of implementing this is picked?

Darksonn commented:
I would rather make these methods depend on `blocking` than make it a dependency of `sync`.

xd009642 commented:
@Darksonn I think this is ready for another review now. I haven't worked to do the deduplication mentioned in review comments yet because I wasn't sure about getting around the moves, though I haven't tried to solve it while sorting out the runtime::enter stuff.

Darksonn commented:
What things are missing when the `blocking` feature flag is not enabled?

xd009642 commented:
When blocking isn't enabled there's no `crate::park::{CachedParkThread, Park};` - or `blocking_send`/`blocking_recv`

carllerche commented on tokio/src/sync/mpsc/bounded.rs:
````diff
@@ -150,6 +152,43 @@ impl<T> Receiver<T> {
+    ///     sync_code.join().unwrap()
+    /// }
+    /// ```
+    #[cfg(feature = "blocking")]
````
I don't think you need a feature flag cfg here. The `blocking` cfg flag is just for enabling the runtime's blocking pool.

carllerche commented on tokio/src/sync/mpsc/bounded.rs:
```diff
@@ -150,6 +152,43 @@ impl<T> Receiver<T> {
+    #[cfg(feature = "blocking")]
+    pub fn blocking_recv(&mut self) -> Result<Option<T>, ()> {
+        let _enter_handle = crate::runtime::enter::enter(false);
+        let mut park = CachedParkThread::new();
```
Instead of implementing the blocking logic, you should be able to use `enter_handle.block_on(&self.recv())`.

You will need to add `sync` to [this](https://github.com/tokio-rs/tokio/blob/646fbae76535e397ef79dbcaacb945d4c829f666/tokio/src/macros/cfg.rs#L59-L67)

xd009642 replied:
I guess I should also use block_on for the `blocking_send` implementation as well?

xd009642 replied:
@carllerche using block_on broke the features ci run, so I added a `cfg_block_on` around both methods. If it's still not working or there's any other comments I'll be around tomorrow morning to resolve them :smile:

xd009642 commented:
@Darksonn @carllerche I think I've addressed all the review comments and all the pending CI is things that have never failed for me so I should be ready another review 😁

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
```diff
@@ -477,3 +507,32 @@ impl<T> Sender<T> {
     }
 }
+
+cfg_block_on! {
+    impl<T> Sender<T> {
```
In this specific case I prefer to manually write the `#[cfg(...)]` instead of using our `cfg_*` macros. This is because having separate impl blocks clutters the documentation for the channel.

Can you move it into the existing impl block between `send_timeout` and `poll_ready`? You can see on the `send_timeout` method how to do it manually without the `cfg_*` macro.

```rust
#[cfg(feature = "time")]
#[cfg_attr(docsrs, doc(cfg(feature = "time")))]
pub async fn send_timeout(
    &mut self,
    value: T,
    timeout: Duration,
) -> Result<(), SendTimeoutError<T>> {
```

xd009642 replied:
Will do, I should get to it in the next 2-3 hours :+1:

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
````diff
@@ -155,6 +155,38 @@ impl<T> Receiver<T> {
+    ///     sync_code.join().unwrap()
+    /// }
+    /// ```
+    #[cfg(any(feature = "blocking", feature = "rt-core", feature = "sync"))]
````
The bounded channel is defined in a module that requires the `sync` feature, so `any(other_stuff, feature = "sync")` should always be satisifed?

xd009642 replied:
I tried without "sync" thinking that and it resulted in some unused item warnings in other modules that made the features CI job fail. The only way I could get rid of those warnings was by adding it

Darksonn replied:
Was that before you added `feature = "sync"` to `cfg_block_on`? I think you should give it another try now.

xd009642 replied:
That was the last failing commit 08197ebe9, so sync was in cfg_block_on

Darksonn replied:
Oh, the cfgs here is not what the error is about. It's due to warnings about things not being in use when only `sync` is enabled. Try removing the `feature = "sync"` clause from `cfg_block_on` and then change this function to
```rust
#[cfg(feature = "rt-core")]
#[cfg_attr(docsrs, doc(cfg(feature = "rt-core")))]
pub fn blocking_recv(&mut self) -> Option<T> {
```
I'm pretty sure this should work. Adding the other things from `cfg_block_on` is redundant, because if you look in [our `Cargo.toml`](https://github.com/tokio-rs/tokio/blob/1167c09ae8bfd0d4fff97d33803d959222f9e2ab/tokio/Cargo.toml#L50-L91), every other feature in the list requires `rt-core`.

Darksonn commented on tokio/src/sync/mpsc/bounded.rs:
```diff
@@ -157,6 +157,13 @@ impl<T> Receiver<T> {
     ///
+    /// # Panics
+    ///
+    /// This function panics if the provided future panics, or if called
+    /// within an asynchronous execution context.
```
The user doesn't provide any future here, so we shouldn't be talking about a "provided" future.

xd009642 replied:
Sorted :+1:
